armadillo/src/main/java/org/molgenis/armadillo/exceptions/InvalidProjectNameException.java

```diff
@@ -0,0 +1,18 @@
+package org.molgenis.armadillo.exceptions;
+
+import static java.lang.String.format;
+import static org.springframework.http.HttpStatus.BAD_REQUEST;
+
+import org.springframework.web.bind.annotation.ResponseStatus;
+
+@ResponseStatus(BAD_REQUEST)
+public class InvalidProjectNameException extends RuntimeException {
+
+  public InvalidProjectNameException(String projectName) {
+    super(
+        format(
+            "Project name '%s' is invalid. Names should adhere to the rules described at "
```

Comment:
I don't mind that we do it, but I just want to note that me that creating classes for each variant of exception look a like a bit 'code heavy' in my eyes. Rather I would prefer a few high level exceptions that are parameterized with the right message (and if you really want to standardize those message to use a central repository of messages). But I do understand this is a 'perfect pattern' that works well. (you are probably irritated by the lack of diversity in exceptions in emx2)

Reply:
I am very very in favour of this method and I actually think this is one of the things MOLGENIS got right 😉 . Here's why:

- It improves readability because the code is not littered with messages
- It simplifies writing exception handling because there is a list of exceptions to choose from
- It ensures consistency of error messages because people don't need to think of one every time they throw an exception
- The exceptions can actually be acted on: for example, we return a `409` on `DuplicateProjectException`, and a `404` on `UnknownProjectException`. This benefit is lost with general exceptions.
- It doesn't take more than a minute to create a new exception
- I would actually consider this method less 'code heavy', because you don't pollute the actual code with long and often duplicated strings

Some of these problems can also be solved by a central repository but then people still are free to use whatever message they want (which in my eyes is a bad thing). If you want to support i18n in the future, passing strings to exceptions will be a pain to refactor (as we have experienced with MOLGENIS 😬).